src/Umbraco.Web/Unversion/UnversionService.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Umbraco.Core;
+using Umbraco.Core.Logging;
+using Umbraco.Core.Models;
+using Umbraco.Core.Services;
+
+namespace Umbraco.Web.Unversion
+{
+    public class UnversionService : IUnversionService
+    {
+        private readonly ILogger _logger;
+        private readonly IUmbracoContextFactory _context;
+        private readonly IContentService _contentService;
+        private IUnversionConfig _config;
+
+        // WB: Note do we want this service in Current.Services etc... ?!
+        public UnversionService(IUnversionConfig config, ILogger logger, IUmbracoContextFactory context, IContentService contentService)
+        {
+            _logger = logger;
+            _config = config;
+            _context = context;
+            _contentService = contentService;
+        }
+
+        public void Unversion(IContent content)
+        {
+            // Get configuration entries for this piece of content
+            var configEntries = new List<UnversionConfigEntry>();
+
+            // Add all config entries if it matches our ContentType Alias
+            if (_config.ConfigEntries.ContainsKey(content.ContentType.Alias))
+                configEntries.AddRange(_config.ConfigEntries[content.ContentType.Alias]);
+
+            // Add all config entries (would it be more than one though?)
+            if (_config.ConfigEntries.ContainsKey(UnversionConfig.AllDocumentTypesKey))
+                configEntries.AddRange(_config.ConfigEntries[UnversionConfig.AllDocumentTypesKey]);
+
+            if (configEntries.Count <= 0)
+            {
+                // Would this ever get hit ?
+                _logger.Debug<UnversionService>("No unversion configuration found for type {ContentTypeAlias}", content.ContentType.Alias);
+                return;
+            }
+
+            foreach (var configEntry in configEntries)
+            {
+                var isValid = true;
+
+                // Check if RootXPath is configured
+                if (!string.IsNullOrEmpty(configEntry.RootXPath))
+                    if (content.Level > 1 && content.ParentId > 0)
+                    {
+                        var ids = GetNodeIdsFromXpath(configEntry.RootXPath);
+                        isValid = ids.Contains(content.ParentId);
+                    }
+
+                if (!isValid)
+                {
+                    _logger.Debug<UnversionService>("Invalid RootXPath configuration for '{ContentTypeAlias}' the XPath selector does not contain the parent id '{ContentParentId}' for the current content item '{ContentUdi}'.",
+                        configEntry.DocTypeAlias, content.ParentId, content.GetUdi());
+
+                    continue;
+                }
+
+                // Get every version of content node
+                var allVersions = _contentService.GetVersionsSlim(content.Id, 0, int.MaxValue).ToList();
```

Comment:
This needs to be paged, this could be enormous. Like on slack, we should only process smaller batches each time until eventually they are all gone.

Reply:
We should not be processing one single item at a time, this is going to kill performance, it all needs to happen in bulk.